src/main/utility/mod.rs

```diff
@@ -19,23 +19,75 @@ pub mod stream_len;
 pub mod syscall;
 pub mod time;
 
+use crate::{core::worker::Worker, host::host::HostId};
+
 /// A type that allows us to make a pointer Send + Sync since there is no way
 /// to add these traits to the pointer itself.
 #[derive(Clone, Copy, Debug)]
-pub struct SyncSendPointer<T>(pub *mut T);
+pub struct SyncSendPointer<T>(*mut T);
 
 unsafe impl<T> Send for SyncSendPointer<T> {}
 unsafe impl<T> Sync for SyncSendPointer<T> {}
 
 impl<T> SyncSendPointer<T> {
-    /// Get the pointer.
+    /// SAFETY: The object pointed to by `ptr` must actually be `Send` and
+    /// `Sync`, or else not subsequently used in contexts where it matters.
+    pub unsafe fn new(ptr: *mut T) -> Self {
+        Self(ptr)
+    }
+
     pub fn ptr(&self) -> *mut T {
         self.0
     }
+}
+
+/// A type that allows us to make a pointer Send + Sync since there is no way
+/// to add these traits to the pointer itself.
+#[derive(Clone, Copy, Debug)]
+pub struct HostTreePointer<T> {
+    #[cfg(debug_assertions)]
+    host_id: HostId,
+    ptr: *mut T,
+}
+
+unsafe impl<T> Send for HostTreePointer<T> {}
+unsafe impl<T> Sync for HostTreePointer<T> {}
+
+impl<T> HostTreePointer<T> {
+    /// SAFETY: The object pointed to by `ptr` must either actually be `Send`
+    /// and `Sync`, or else not subsequently used in contexts where it matters.
+    ///
+    /// In Shadow typically the object itself may not be `Send` or `Sync`, but
+    /// safety is guaranteed by the pointer living within a single Host's
+    /// "object tree", and accessed with that Host's lock held.
+    pub unsafe fn new_for_host(host_id: HostId, ptr: *mut T) -> Self {
+        // Suppress "never read" warning in release builds.
+        let _ = host_id;
+        Self {
+            #[cfg(debug_assertions)]
+            host_id,
+            ptr,
+        }
+    }
+
+    /// Create a pointer for the current host.
+    ///
+    /// SAFETY: see `new_for_host`.
+    pub unsafe fn new(ptr: *mut T) -> Self {
```

Comment:
Does this need to be `unsafe`? The `Worker::with_active_host_info()` below is already checking for the host lock and binding this to that specific host, so there aren't any other guarantees that the calling code needs to uphold.

Reply:
My thinking *was* that the runtime checks would only be enabled in debug builds, as for the Magic checks. The debug checks are a best effort to catch mistakes, but it's still the user's responsibility to avoid mistakes.

On further thought and after having worked out the more complete design of this idea, I think I agree though that it's better to think of these checks more like those in a RefCell and keep them on in release builds. Taking another pass...

Reply:
Done - this is no longer `unsafe`